framework/src/meshgenerators/FancyExtruderGenerator.C

```diff
@@ -0,0 +1,547 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FancyExtruderGenerator.h"
+
+#include "libmesh/boundary_info.h"
+#include "libmesh/function_base.h"
+#include "libmesh/cell_tet4.h"
+#include "libmesh/cell_tet10.h"
+#include "libmesh/face_tri3.h"
+#include "libmesh/face_tri6.h"
+#include "libmesh/face_quad4.h"
+#include "libmesh/face_quad9.h"
+#include "libmesh/cell_prism6.h"
+#include "libmesh/cell_prism18.h"
+#include "libmesh/cell_hex8.h"
+#include "libmesh/cell_hex27.h"
+#include "libmesh/libmesh_logging.h"
+#include "libmesh/mesh_communication.h"
+#include "libmesh/mesh_modification.h"
+#include "libmesh/mesh_tools.h"
+#include "libmesh/parallel.h"
+#include "libmesh/remote_elem.h"
+#include "libmesh/string_to_enum.h"
+#include "libmesh/unstructured_mesh.h"
+
+#include <numeric>
+
+registerMooseObject("MooseApp", FancyExtruderGenerator);
+
+template <>
+InputParameters
+validParams<FancyExtruderGenerator>()
+{
+  InputParameters params = validParams<MeshGenerator>();
+
+  params.addRequiredParam<MeshGeneratorName>("input", "The mesh to extrude");
+
+  params.addClassDescription("Extrudes a 2D mesh into 3D, can have variable a variable height for "
+                             "each elevation, variable number of layers within each elevation and "
+                             "remap subdomain_ids within each elevation");
+
+  params.addRequiredParam<std::vector<Real>>("heights", "The height of each elevation");
+
+  params.addRequiredParam<std::vector<unsigned int>>(
+      "num_layers", "The number of layers for each elevation - must be num_elevations in length!");
+
+  params.addParam<std::vector<std::vector<subdomain_id_type>>>(
+      "subdomain_swaps",
+      "For each row, every two entries are interpreted as a pair of "
+      "'from' and 'to' to remap the subdomains for that elevation");
+
+  params.addRequiredParam<Point>(
+      "direction",
+      "A vector that points in the direction to extrude (note, this will be "
+      "normalized internally - so don't worry about it here)");
+
+  params.addParam<boundary_id_type>(
+      "top_boundary",
+      "The boundary ID to set on the top boundary.  If ommitted one will be generated.");
+
+  params.addParam<boundary_id_type>(
+      "bottom_boundary",
+      "The boundary ID to set on the bottom boundary.  If omitted one will be generated.");
+
+  return params;
+}
+
+FancyExtruderGenerator::FancyExtruderGenerator(const InputParameters & parameters)
```

Comment:
Do we want to call it *Fancy*? Maybe just name it as `ExtruderGenerator` and get rid of the old one. In the future, we may just rename it `Extruder`. I do not think there is anything the old one can do but this new one cannot.

Reply:
I am guessing only moose is using that old extruder generator (not the modifier), so getting rid of it should be possible.

Reply:
Yeah - Fancy is actually a pretty non-descriptive name for this object. Maybe `AdvancedExtruder` or `FlexibleExtruder`, or `VariableLayerExtruder`. If we get rid of the old one than just `Extruder` would be nice too. I suppose we can just default to generating like the old one? It should be difficult to deprecate and remove the existing object.

Reply:
Well - it's just "fancy".  It's pretty clear that this is most-likely an intermediate name anyway...